### PR TITLE
RUMM-2172 Single Storage

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Provide an implementation of the single storage mechanism for SDK v2

### Additional Notes

Unit tests are globally not passing but all tests on the new classes are green ✅

### CODE REVIEW

**Review by 0xnm — APPROVED** [review]
Nicely done! I've added some thoughts/comments, but none of them are blocking for me.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            }
+
+            override fun write(event: ByteArray, eventId: String, newMetadata: ByteArray) {
+                // prevent useless operation for empty event / null orchestrator
+                if (event.isEmpty() || orchestrator == null) {
+                    listener.onDataWritten(eventId)
+                    return
+                }
+
+                val file = orchestrator.getWritableFile(event.size)
```
[the remark this answers is not in the document]
Comment (ncreated) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt:55]: > (...) by the time it is invoked we may have other consent -> different orchestrator, not the one we captured for invocation

Not sure if I understand this @0xnm 🤔💭. We get the orchestrator based on the immutable `datadogContext` passed as an parameter to `writeCurrentBatch()`. Then, any time later we receive a callback which provides us API (`BatchWriter`) to writing data. The write strategy will respect the consent value from `datadogContext`.

From the caller's POV, IMO everything should be consistent with this implementation: the caller orders write for given consent value and it knows that `BatchWriter` will respect it. No?

**Review by ncreated — APPROVED** [review]
Looks very good from the architecture POV 👍. I can see slight differences when comparing to RFC interfaces, but I believe these are Android / Kotlin specifics.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt:55]: I mean that generally the API contract is that this callback may be invoked in async manner (if I got it right). My point is that what if user changed a consent in the timeframe between the moment we received an `orchestrator` and the moment we use it in the `BatchWriter`. Then event may be written to the location which belongs to the old consent.

This is quite hypothetical case (and may be quite an edge case), but this gist of it is that imo we need to keep in mind the possible SDK state changes with async invocations.

File: dd-sdk-android/src/main/kotlin/com/datadog/android/v2/api/context/TimeInfo.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.v2.api.context
+
+/**
+ * Holds information about the current local and server time.
+ * @property deviceTimeNs the current time as known by the System on the device (nanoseconds)
+ * @property serverTimeNs the current time synchronized with our NTP server(s) (nanoseconds)
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/v2/api/context/TimeInfo.kt:12]: 👍

File: dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (xgouchet) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt:55]: This is an edge case that is not something we want to dig into. The time frame for such a concurrency issue will on most case something smaller than 500ms, which is totally acceptable.

File: dd-sdk-android/src/test/kotlin/com/datadog/android/utils/forge/SDKContextForgeryFactory.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        return SDKContext(
+            TimeInfo(forge.aLong(), forge.aLong()),
+            ApplicationInfo(
+                forge.anAlphabeticalString(),
+                forge.anAlphabeticalString(),
+                forge.anInt()
+            ),
+            ProcessInfo(forge.aBool(), forge.anInt()),
+            NetworkInfo(forge.aValueFrom(NetworkInfo.Connectivity::class.java), null),
+            UserInfo(null, null, null, emptyMap()),
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/utils/forge/SDKContextForgeryFactory.kt:33]: Yes I know, I'll deal with those in a future PR (when those are actually needed)

File: dd-sdk-android/src/test/kotlin/com/datadog/android/v2/core/DatadogCoreInitializationTest.kt [hunk 4]
```diff
@@ -119,13 +94,13 @@ internal class DatadogCoreInitializationTest {
...
 
         // Then
-        assertThat(CoreFeature.initialized.get()).isTrue()
-        assertThat(LogsFeature.initialized.get()).isEqualTo(logsEnabled)
-        assertThat(TracingFeature.initialized.get()).isEqualTo(tracingEnabled)
-        assertThat(CrashReportsFeature.initialized.get()).isEqualTo(crashReportEnabled)
-        assertThat(RumFeature.initialized.get()).isEqualTo(rumEnabled)
-        assertThat(WebViewLogsFeature.initialized.get()).isEqualTo(logsEnabled)
-        assertThat(WebViewRumFeature.initialized.get()).isEqualTo(rumEnabled)
+        // assertThat(CoreFeature.initialized.get()).isTrue()
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply dd-sdk-android/src/test/kotlin/com/datadog/android/v2/core/DatadogCoreInitializationTest.kt:97]: Yes good point

File: dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply dd-sdk-android/src/main/kotlin/com/datadog/android/v2/core/internal/storage/ConsentAwareStorage.kt:55]: I get the point now @0xnm 👌. Agree, it's kind of edge case, but it can be clearly visible in such user situation:
```swift
// all synchronous, on a single thread:
Datadog.set(trackingConsent: .granted)
logger.info("granted log")
Datadog.set(trackingConsent: .pending)
```
In such case, the _"granted log"_ should be sent by the SDK. This is however a question on how we will further synchronise things in higher implementation (even higher than `Storage` level). The RFC recommends processing all user calls through context thread first (see [_Threading in `DDCore`_](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/2344748708/RFC+-+V2+Core+Storage+and+Batching#Threading-in-DDCore)) to support exactly these edge cases.
